### PR TITLE
feat: add support for clustered Redis caches

### PR DESCRIPTION
**Pull request by neoeinstein**
## Changes

Adds a new global configuration option with supporting documentation that allows connections to Redis clusters, enabling their use for caching.

## Context

Implements discussion [#31884](https://github.com/renovatebot/renovate/discussions/31184)

## Documentation (please check one with an [x])

- [X] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [X] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — CHANGES_REQUESTED** [review]
I would prefer to encode it into the url, like other tools. so we don't need an extra config option.

eg `redis-cluster://....`

**Comment by neoeinstein** [comment]
Note that this would then require (some minimal) parsing and then re-writing the URLs to set the schema back to `redis` or `rediss` as appropriate in order to accomplish this, and would not be in the format generally expected for Redis URLs. That's probably fine if documented, but I want to make sure that's the explicit decision here.

In addition, if we go this way, I'd recommend `redis+cluster` as that seems to be the general convention used when specifying a "protocol" extension in the schema.

While not included here, using the mode parameter instead would make it easier to accept and map multiple root nodes by allowing `redisUrl` to be a comma-separated list of root nodes without ambiguity.

**Comment by viceice** [comment]
> Note that this would then require (some minimal) parsing and then re-writing the URLs to set the schema back to `redis` or `rediss` as appropriate in order to accomplish this, and would not be in the format generally expected for Redis URLs. That's probably fine if documented, but I want to make sure that's the explicit decision here.
> 
> In addition, if we go this way, I'd recommend `redis+cluster` as that seems to be the general convention used when specifying a "protocol" extension in the schema.
> 
> While not included here, using the mode parameter instead would make it easier to accept and map multiple root nodes by allowing `redisUrl` to be a comma-separated list of root nodes without ambiguity.

I would like to find a unified configuration for the package cache, so we can combine it with the SQLite[^1] cache option.
If not defined, use the current default cacache implementation.

[^1]: https://docs.renovatebot.com/self-hosted-experimental/#renovate_x_sqlite_package_cache

**Comment by rarkins** [comment]
Here's how keyv handles Redis clusters: https://keyv.org/docs/storage-adapters/redis/

@viceice are you hoping for a single URL to define a package cache completely? e.g. `sqlite://` `redis://` `rediss://` `rediss+cluster://` etc?

**Comment by viceice** [comment]
> Here's how keyv handles Redis clusters: https://keyv.org/docs/storage-adapters/redis/
> 
> @viceice are you hoping for a single URL to define a package cache completely? e.g. `sqlite://` `redis://` `rediss://` `rediss+cluster://` etc?

yes, something like that. ❤️

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/cache/package/redis.ts [hunk 1]
```diff
@@ -94,16 +97,28 @@ export async function init(
   }
   rprefix = prefix ?? '';
   logger.debug('Redis cache init');
-  client = createClient({
-    url,
+
+  const rewrittenUrl = url.replace(/^(redis|rediss)\+cluster:\/\//, '$1://');
```
Comment (viceice) [inline lib/util/cache/package/redis.ts:101]:
```suggestion
  const rewrittenUrl = url.replace(regEx(/^rediss?\+cluster:\/\//), '$1://');
```

**Review by viceice — COMMENTED** [review]
this needs some unit tests

File: lib/util/cache/package/redis.ts (hunk 1 above)
Comment (neoeinstein) [reply lib/util/cache/package/redis.ts]: Updated to add unit tests around the application, used the `regEx` internal type, and simplified the RegEx.

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**